AIChat full-page UI

petemill opened the pull request:
Resolves https://github.com/brave/brave-browser/issues/30554

Introduces:
- Full-page conversation UI at chrome-untrusted://chat
- Selectable older conversations in full-page mode
- Selectable older conversations in sidebar mode
- Browser command to open full-page mode (exposed to _Quick Commands_ and _Shortcuts_)

_note: all functionality is behind the feature flag: `AIChatHistory`_

## TODO:
- [x] "Expand" button from sidebar to full-page should be on the header, not in the kebab menu
- [x] Links between sidebar and full-page modes
- [x] Restrict access to full-page if the AIChatHistory flag isn't enabled
  - [x] Bring back the AIChatThrottle
- [x] Revert toolbar button to open sidebar
- [x] When page is narrow, floating navigation panel should start collapsed
- [x] Eraser and + buttons should both be pencil 

### UI Specifics:
https://www.figma.com/design/m0Gdbf0wtqyfEFGm32VLLc/%F0%9F%94%84-Leo-%5BIN-PROGRESS%5D?node-id=15-15519&node-type=canvas&t=2V4x2M8hF3ViLVji-0
- [x] conversation title should show:
    - [x] on full-page: always
    - [x] on sidebar: only when an archive conversation is being viewed
- [x] header bottom border should line up between navigation and main areas
- [x] Merge kebab menus to a single menu

### Follow-up
- [ ] Outer trusted webui and inner frame untrusted webui for conversation content
- [ ] If storage flag is disabled then full-page should display info about enabling the flag (or preference should that exist)
- [ ] Loading indicator / skeleton for conversation list (and conversation entries)
- [ ] Rename `chat` hostname to `leo`
- [ ] Titles for conversations for BYOM
- [ ] Url path for conversation ID https://github.com/brave/brave-core/pull/26050
  - [x] Remove back button from full-page
  - [ ] Make omnibox item open full-page
- [ ] Delete confirmation dialog (can be follow-up)
- [ ] Mobile navigation / header variations

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

petemill commented on components/ai_chat/resources/page/components/feature_button_menu/index.tsx:
```diff
@@ -28,11 +28,13 @@ export default function FeatureMenu(props: Props) {
     aiChatContext.uiHandler?.openAIChatSettings()
   }
 
+  const inConversation = conversationContext.conversationHistory.length >= 1
```
We'll need to think about this. I assume the intention here is to work out if we're in a "deletable" and "renameable" conversation? The only way to know this is to see if the conversation is in the list of visible conversations returned by the Service. A conversation can have chat messages but not be in that list, e.g. when it has staged content from Brave Search.

fallaciousreasoning commented:
Sorry @petemill it wasn't building when I switched back after the policy stuff landed, so I rebased

petemill commented:
@fallaciousreasoning feel free to rebase whenever neccessary!

fallaciousreasoning commented on components/ai_chat/resources/page/state/conversation_context.tsx:
```diff
@@ -282,6 +282,12 @@ export function ConversationContextProvider(
 
+    id = callbackRouter.onConversationDeleted.addListener(() => {
+      // TODO(petemill): Show deleted UI
+      console.debug('DELETED')
```
@petemill do we need to do something more here?

petemill replied:
I was leaving it open for us to display a note, something like "this conversation has been deleted", or switch to just switch to a new conversation

fallaciousreasoning commented on components/ai_chat/resources/page/chat_ui.tsx:
```diff
@@ -83,21 +86,35 @@ function App() {
+    return <div>"loading..."</div>
+  }
+
+  if (!aiChatContext.isStandalone) {
```
@petemill I wonder if this would be a better match for `loadTimeData` rather than an asynchronous API call, as it means we'll block loading this page every time until we've:
1. Loaded the page
2. Loaded/parsed/executed the JS
3. Done a React render
4. Done a mojom round trip to the browser to determine whether we're in `isStandalone` mode
5. Done a React rerender

petemill replied:
Yes as long as the timing works out for the compaison of the web_ui WebContents to the active WebContents to be done in the constructor.

petemill commented on app/brave_command_ids.h:
```diff
@@ -137,6 +137,8 @@
 // Adblock (user-defined cosmetic filters)
 #define IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS 56351
 
+#define IDC_OPEN_FULL_PAGE_CHAT 56352
```
@fallaciousreasoning should this be grouped with the other AI_CHAT commands above?

petemill commented on browser/ui/brave_pages.cc:
```diff
@@ -45,6 +45,10 @@ void ShowBraveTalk(Browser* browser) {
 }
 
+void ShowFullpageChat(Browser* browser) {
+  ShowSingletonTabOverwritingNTP(browser, GURL("chrome-untrusted://chat"));
```
TODO: only do this if ai_chat::features::IsAIChatHistoryEnabled()

petemill commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -217,6 +221,36 @@ void AIChatService::DismissPremiumPrompt() {
+  if (!conversation_handler) {
+    return;
+  }
+  // conversation_handler->OnConversationDeleted();
```
TODO: see if we can make the timing of this work (notify UI clients of the conversation)

fallaciousreasoning replied:
Do we need to solve this before landing?

petemill replied:
Not needed as the call to `OnConversationDeleted` in the destructor works anyway

petemill commented on browser/ui/brave_browser_command_controller.cc:
```diff
@@ -360,6 +360,7 @@ void BraveBrowserCommandController::UpdateCommandForAIChat() {
   bool command_enabled = (sidebar::CanUseSidebar(&*browser_) &&
                           ai_chat::IsAllowedForContext(browser_->profile()));
   UpdateCommandEnabled(IDC_TOGGLE_AI_CHAT, command_enabled);
+  UpdateCommandEnabled(IDC_OPEN_FULL_PAGE_CHAT, true);
```
should only be enabled if AIChatHistory feature flag is enabled and AI Chat is allowed in the context

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-fullpage/storybook.html)

fallaciousreasoning commented on components/ai_chat/resources/page/components/sidebar_nav/index.tsx:
```diff
@@ -0,0 +1,186 @@
+        )}
+        <form className={styles.form}>
+          <input
+            placeholder='Search your conversations'
```
nit: translate

petemill replied:
oh actually we need to completely remove search

fallaciousreasoning commented on components/ai_chat/resources/page/components/sidebar_nav/index.tsx:
```diff
@@ -0,0 +1,186 @@
+                    ) : (
+                      <DisplayTitle
+                        title={item.title || getLocale('conversationListUntitled')}
+                        description={''}
```
```suggestion
                        description=''
```

fallaciousreasoning commented on components/ai_chat/resources/page/components/main/index.tsx:
```diff
@@ -159,44 +159,18 @@ function Main() {
+      {isConversationListOpen && (
+        <div className={styles.conversationList}>
+          <SidebarNav
+            enableBackButton={true}
```
nit:
```suggestion
            enableBackButton
```

petemill replied:
I removed this variable!

fallaciousreasoning commented on components/ai_chat/resources/page/components/full_screen/index.tsx:
```diff
@@ -0,0 +1,129 @@
+  const [isOpen, setIsOpen] = React.useState(isSmall)
+  const [shouldRender, setShouldRender] = React.useState(!isSmall)
+
+  const initAsideAnimation = React.useCallback((node: HTMLElement | null) => {
+    if (!node) return
+    const open = { width: '340px', opacity: 1 }
+    const close = { width: '0px', opacity: 0 }
+    const animationOptions: KeyframeAnimationOptions = {
+      duration: 200,
+      easing: 'ease-out',
+      fill: 'forwards'
+    }
+    asideAnimationRef.current = new Animation(
+      new KeyframeEffect(node, [open, close], animationOptions)
+    )
+
+    // Make sure we're in the right state for our screen size when
+    asideAnimationRef.current.playbackRate = isSmall ? 1 : -1
+    asideAnimationRef.current.finish()
+  }, [])
```
@petemill in a followup I'd like to revisit how the sidebar works, because its kinda painful to use

petemill replied:
> how ~the sidebar~ navigation works

👍

fallaciousreasoning commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -354,6 +401,8 @@ std::vector<mojom::Conversation*> AIChatService::FilterVisibleConversations() {
     conversations.push_back(conversation.get());
   }
+  base::ranges::sort(conversations, std::greater<>(),
+                     &mojom::Conversation::created_time);
   return conversations;
```
wow, cool!

fallaciousreasoning reviewed (commented):
This is looking pretty great, if I do say so myself

kdenhartog commented on components/ai_chat/resources/page/ai_chat_ui.html:
```diff
@@ -12,6 +12,7 @@
     <script src="//resources/js/load_time_data_deprecated.js"></script>
     <script src="/strings.js"></script>
     <link rel="stylesheet" href="./styles.css">
+    <link rel="icon" href="//resources/brave-icons/social-leo-favicon-fullheight-color.svg">
```
No concerns to me - looks like these are built in.

petemill commented:
@fallaciousreasoning I've reviewed yours (and @nullhook 's original changes, and my refinements!) and I "approve", although Github won't let me stamp it accordingly because I created the PR.

I've also pushed some file, component and variable renames, deleted code not used in this version, and cleaned up the conversation titles display.

Please review those last commits @fallaciousreasoning. Also cc @Brandon-T for the minor iOS function stubs where I don't think we need to do anything yet.

bsclifton reviewed (approved):
String reviewers ++

brave-builds commented:
Released in [v1.73.62](https://github.com/brave/brave-browser/releases/tag/v1.73.62)
